Add tests for custom JSON-RPC network client

BelfordZ opened the pull request:
## Explanation

<!--
Thanks for the pull request. Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?

Are there any issues, Slack conversations, Zendesk issues, user stories, etc. reviewers should consult to understand this pull request better? For instance:

* Fixes #12345
* See: #67890
-->

Previously we had written tests for `createInfuraClient`, which creates a middleware stack designed to connect to an Infura provider. These tests exercise various RPC methods that relate to the behavior that the middleware provides (mainly around caching). 

Now we need to write the same tests but for `createJsonRpcClient`, which creates a middleware stack designed to connect to a non-Infura RPC endpoint. To do this, we had to:

- Consolidate the tests for both types of RPC client into a single test file.
- Add conditions around tests or assertions in tests to account for differences in behavior between the two sets of middleware stacks.
- Relocate code in `createJsonRpcClient` which slows down `eth_estimateGas` calls just for tests so that this behavior can be disabled in the network client tests

It's probably best to look at `shared-tests.js` to get a better sense of the tests that now exist between the two network clients. We may introduce PRs in the future to make this easier to read.

Eventually, as we unify the network controllers in this repo and in the controllers repo, we will move these tests into the controllers repo.

## Manual Testing Steps

<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

This PR does not change any user-facing behavior. All of the tests should continue to pass with this PR.

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

BelfordZ committed:
Add shared network tests to createJsonRPCClient

Changes the original shared-tests & underlying helpers for network
middleware testing. Also Removes test middleware from createJsonRpcClient

BelfordZ commented:
Removed the code for estimate gas in test, everything still green. :) :)

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/helpers.js:
```diff
@@ -286,14 +295,23 @@ export async function withMockedInfuraCommunications(...args) {
+  } else {
+    clientUnderTest = createJsonRpcClient({
+      rpcUrl: 'http://localhost:8545',
+      chain_id: '0x1',
```
another note, kind of surprising that this doesn't make some tests fail (some tests are using different networks and checking their chain id...)

BelfordZ replied:
should this be `chainId`

mcmire replied:
Hmm. I'll take a closer look at this.

BelfordZ replied:
Ive fixed this issue actually. There was some code that wasnt passing in type. This would have been caught if it was required, so + points to that suggestion :)

mcmire replied:
Oh ok cool!

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/helpers.js:
```diff
@@ -286,14 +295,23 @@ export async function withMockedInfuraCommunications(...args) {
  * on that object.
  * @returns {Promise<any>} The return value of the given function.
```
todo: update jsdoc

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/helpers.js:
```diff
@@ -249,21 +258,21 @@ function makeRpcCall(ethQuery, request) {
  * is called with an object that allows you to mock different kinds of requests.
```
todo: update jsdoc

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/helpers.js:
```diff
@@ -94,12 +95,24 @@ function debug(...args) {
- * @param {string} options.network - The Infura network you're testing with
+ * @param {string} options.network - The network you're testing with
  * (default: "mainnet").
+ * @param {string} options.type - Infura or Custom
```
```suggestion
 * @param {string} options.type - if defined, must be either `infura` or `custom`
```

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/helpers.js:
```diff
@@ -194,7 +200,10 @@
     }
   }
-  let nockRequest = nockScope.post(`/v3/${INFURA_PROJECT_ID}`, {
+  const url = nockScope.basePath.includes('infura.io')
```
CodeQL must not know this is only for testing. hmm.

BelfordZ replied:
the url looks more like https://${network}.infura.io

It's right that its a lazy way to check a url, but its wrong about the context and how much it should matter. If it knew it was test code, we wouldnt have got this error.

mcmire replied:
Ah yeah I got the URL wrong, lol. But yeah I see what you mean about the CodeQL check. No reason it should be here, and no reason it should be a High at the very least.

BelfordZ commented on app/scripts/controllers/network/createJsonRpcClient.test.js:
```diff
@@ -0,0 +1,194 @@
+/**
```
same tests suite as createInfuraClient, just moving it closer towards using the openrpc document referenced in the comment below

BelfordZ replied:
I will add that back. 

At some point I'd like to revist these tests and use a combination of the ethereum/execution-apis + metamask/api-specs openrpc document

BelfordZ commented on app/scripts/controllers/network/createJsonRpcClient.test.js:
```diff
@@ -0,0 +1,194 @@
+    { name: 'eth_unsubscribe', numberOfParameters: 1 },
+    { name: 'net_listening', numberOfParameters: 0 },
+    { name: 'net_peerCount', numberOfParameters: 0 },
+    { name: 'parity_nextNonce', numberOfParameters: 1 },
```
`parity_nextNonce` - I tested this method on a new infura api key and seems it doesn't work. Do we need to include this method, or any method with `parity` prefix?

mcmire replied:
This is why I differentiated what was in Infura but not in the spec, or vice versa, because we seem to have legacy code in the middleware. So if we are explicitly handling this, I think we should test it, even if it's not supported in the spec, but we should note as such so that we know that it's legacy and that we should remove it in a future PR.

BelfordZ replied:
ah okay. I did a rough search for things touching parity_nextNonce and came up with nothing, but to be fair I didnt search very exhaustively.

BelfordZ commented on app/scripts/controllers/network/createJsonRpcClient.test.js:
```diff
@@ -0,0 +1,194 @@
+        },
+      );
+
+      expect(chainId).toStrictEqual('5');
```
I would have thought this would fail since we hardcode chain_id: "0x1"

mcmire replied:
Hmm, I'll take a look at this more closely.

BelfordZ replied:
as mentioned above, fixed this now

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -95,58 +117,68 @@ export function testsForRpcMethodNotHandledByMiddleware(
  * parameter will not be used in determining how to cache the method.
  *
  * @param method - The name of the RPC method under test.
+ * @param options0
```
```suggestion
 * @param options the options
```

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -430,12 +488,18 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
+        const errMsg = `${errorMessagePrefix}: Some message`;
+
+        let err;
+        if (providerType === 'infura') {
```
should probably make this into a helper since it appears in a few places

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -447,52 +511,58 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
  * cacheable.
  *
  * @param method - The name of the RPC method under test.
+ * @param options0
```
fix jsdoc

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -511,21 +581,21 @@ export function testsForRpcMethodsThatCheckForBlockHashInResponse(method) {
       const mockResults = [emptyValue, { blockHash: '0x100' }];
 
-      await withMockedInfuraCommunications(async (comms) => {
+      await withMockedCommunications(async (comms) => {
```
missing params here...

Nooknick7899 commented:
2d9a7d6e64258d4c49ed136af9b34fb44187ebb1

BelfordZ committed:
Refactor the createJsonRpcClient & createInfuraClient tests into 1

BelfordZ committed:
Changed custom rpc to return network id for known chain ids

also fixed some tests

mcmire commented on app/scripts/controllers/network/createJsonRpcClient.js:
```diff
@@ -12,21 +12,25 @@ import { SECOND } from '../../../../shared/constants/time';
-};
 
-export default function createJsonRpcClient({ rpcUrl, chainId }) {
+export default function createJsonRpcClient({ rpcUrl, chainId, networkId }) {
```
What do you think about making API/behavior changes to `createJsonRpcClient` in a different PR? That way it might be easier to follow why we are doing this.

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/helpers.js:
```diff
@@ -275,25 +289,35 @@ export async function withMockedInfuraCommunications(...args) {
+  } else {
+    clientUnderTest = createJsonRpcClient({
+      rpcUrl: LOCALHOST_RPC_URL,
+      chainId: BUILT_IN_NETWORKS[network].chainId || '0x1',
```
I have another branch on controllers repo that moves these tests into typescript, and so I will need to adjust this code over there anyways.

mcmire replied:
Okay. I'm not totally convinced we have to do this but perhaps there's something I'm not seeing. Maybe we change this over there in any case so those changes go together?

BelfordZ replied:
Valid points. I thought it would be best to try to keep the scope of change as minimal as possible until we are using network controller from controllers repo.

BelfordZ replied:
That said, I would like to unify the interfaces in the controllers repo PR that is soon to come...

BelfordZ replied:
hmm... Ill give it a shot and see

carlexos replied:
Leave a reaply

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/helpers.js:
```diff
@@ -91,15 +96,27 @@ function debug(...args) {
-  return nock(`https://${network}.infura.io`).filteringRequestBody((body) => {
+function buildScopeForMockingRequests({
+  network = 'mainnet',
+  type = 'infura',
```
The only reason I opted for defaulting it is that it would make for less code to change. Not opposed to the idea of making it required.

BelfordZ replied:
Should I make network also required? Same argument applied right?

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/createClient.test.js:
```diff
@@ -0,0 +1,216 @@
+/**
```
not the biggest fan, because this makes it IMO more difficult to tell what the actual differences are. With this PR, you can search for providerType === 'infura' and find every case where its different. If they are seperate files, over time they will continue to drift and the differences become less clear. I would like us to move towards unifying these 2 interfaces and implementations, and starting with the tests is IMO a good way to get us there. 

One thing that I could get behind is to move all the things that work differently into 2 seperate test files (`createInfuraClient.test.js` and `createJsonRpcClient.test.js`), and leave the things that work the same in a `createClient.test.js`. I'm not sure the juice is worth the squeeze on that one though.

mcmire replied:
> not the biggest fan, because this makes it IMO more difficult to tell what the actual differences are. With this PR, you can search for providerType === 'infura' and find every case where its different.
>
> ...
>
> I would like us to move towards unifying these 2 interfaces and implementations, and starting with the tests is IMO a good way to get us there.

Gotcha. I guess the only other argument I would make here about keeping all the tests in one file vs. two is that it creates more code complexity and that may end up impacting maintainability in the future. I've found, when writing tests, that every time I use loops and logic and other mechanisms that I would usually use in an implementation file to make code more compact, it pretty much always ends up biting me when I go to update those tests later. If you want a straightforward list of what this function does, now you have to go through each test and apply the logic yourself, and it can be a burden to do that.

That said, I see your point about being able to tell the differences more easily if you have the logic right in front of you instead of having to do the comparison between two files yourself. Obviously right now we don't have that much logic in these tests, so it's not a big deal. Also, it sounds like you have a plan for unifying all of this and maybe this will end up getting simplified or something. So, maybe this is just something we watch out for in the future.

BelfordZ replied:
For sure I share in your frustrations of having to deal with overly abstract testing setups, this is also a worry of mine. Theres a delicate tradeoff between additional complexity Vs. how much code needs to be updated for any given change. More places to update leads to greater likelyhood of mistakes being made and not caught. More complexity makes it take longer before someone is able to update / fix any of that code. The list of tradeoffs goes on. Context is important in determining how much of each is appropriate, and in the context of testing, I dont think we should aim to minimize complexity at the cost of extreme duplication, just as I dont think we should aim to optimize for lowering lines of code. We should aim to group functionality into tests that if they break, they are broken in relation to the grouped functionality.

That being said, I think due to the nature of these tests (broad coverage of general functionalities), I think this level of complexity and abstractness is acceptable, and in this case, I think it actually makes them easier to maintain, since changes to the client / middleware require far fewer updates to the tests. I'm certainly not advocating for over-valueing reduction in lines of code, but I believe this to be more of a grouping tests based around some general functionalities. Unless you are changing those general functionalities, you should never end up with one of these tests breaking on you. If you are adding a new method to this set but it doesnt pass, putting it in a different test file makes it clear that there is a deviation from the standard client middleware functionality. 

As for unifying it all, I think we can do this at the createXYZClient interface level to start with at the same time as we migrate this code to controllers repo. This will overall minimize the change set introduced when later integrating controllers/network-controller with extension.

Gudahtt replied:
Hmm. Alternatively, we could take the loop contents here and move them to a "shared client tests" module, and use that in both `createJsonRpcClient.test.js` and `createinfuraClient.test.js`.

It would be nice to keep the convention of the test module matching the name fo the module under test. But in addition to that, I had planned on adding some tests that would only be applicable to the Infura client (for the Infura client specific middleware). That would better fit in a dedicated module for that client as well.

mcmire replied:
> As for unifying it all, I think we can do this at the createXYZClient interface level to start with at the same time as we migrate this code to controllers repo. This will overall minimize the change set introduced when later integrating controllers/network-controller with extension.

Even if we want to unify these interfaces, I am going to push back slightly on this and argue that we should hold off on doing that. Not only would duplicating the tests / test helpers make this PR more focused, as well as the PR that will end up doing the unification, but it also gives us room to change our mind in case the separation does turn out to be useful (as Mark points out).

mcmire replied:
@BelfordZ I pushed a commit which restored `createInfuraClient.test.js` and `createJsonRpcClient.test.js`. The bulk of the tests in `shared-tests.js` as you had them; the individual files just call a function. I did pull out a couple of tests, however, that seemed to be Infura-specific.

BelfordZ committed:
Removed network id cache change

BelfordZ committed:
Revert network controller changes back to original

BelfordZ committed:
Fixed describe block

BelfordZ committed:
Type is required for network client test helpers

BelfordZ committed:
Added chainid as an option to withClient

BelfordZ committed:
Added additional seperation in the network client tests

Gudahtt commented on app/scripts/controllers/network/createJsonRpcClient.js:
```diff
@@ -36,26 +34,3 @@ export default function createJsonRpcClient({ rpcUrl, chainId }) {
   return { networkMiddleware, blockTracker };
 }
-
-function createChainIdMiddleware(chainId) {
```
Could we move these changes to a separate PR?

mcmire replied:
I agree, we shouldn't be making changes to any middleware in this PR, we should just add new tests.

carlexos replied:
Leave a reaply

mcmire replied:
I realized why some of these changes were made and in fact I think it is necessary to adjust this file a bit. It seems that the `eth_gasEstimate` RPC method has been modified, or rather decorated, in tests so that it runs 2 seconds slower. That said I don't think the middleware that is used to achieve this is written correctly, as it seems to pause the request forever. I amended this file in a more minimal way so that we could disable this behavior for the network client tests.

Gudahtt replied:
Interesting. I'm having trouble understanding how your changes here resolved the issue. It appears to be functionally equivalent.

mcmire replied:
This comment is tied to a previous commit, so is it possible you're not seeing the new changes? I can add a comment to the latest commit to explain what's happening.

Gudahtt replied:
Ohh I see. IN_TEST is being changed dynamically. I hadn't considered that, I was used to the e2e context where it's converted to a string literal. But we can change it here.

Gudahtt replied:
Fair enough! This seems like a pretty safe change to make, unlikely to have a functional impact.

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -1108,31 +1166,144 @@ export function testsForRpcMethodSupportingBlockParam(
+      ['ETIMEDOUT', 'ECONNRESET', 'SyntaxError'].forEach(
+        (errorMessagePrefix) => {
+          if (providerType === 'infura' || errorMessagePrefix === 'ETIMEDOUT') {
+            it(`retries the request to Infura up to 5 times if an "${errorMessagePrefix}" error is thrown while making the request, returning the successful result if there is one on the 5th try`, async () => {
```
Hmm, this test description has "Infura" in it, but it can be run with the non-Infura client if the current iteration is for `ETIMEDOUT`

mcmire commented on app/scripts/controllers/network/provider-api-tests/createClient.test.js:
```diff
@@ -0,0 +1,180 @@
+      );
+    });
+
+    it(`net_version does not hit ${providerType} provider, instead returning the networkId that maps to the rpc network, as a decimal string`, async () => {
```
Nit: What do you think about using a `describe` to hold tests for `net_version`, even if there is only one test? This way the names of the tests (when viewed by running this one test file on its own) are consistent, and the test file in general follows a general shape.

mcmire replied:
These tests have been moved to `createInfuraClient.test.js`.

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -447,59 +516,69 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
-  it('does not hit Infura more than once for identical requests and it has a valid blockHash', async () => {
+export function testsForRpcMethodsThatCheckForBlockHashInResponse(
+  method,
+  { providerType = 'infura' } = { providerType: 'infura' },
```
Nit: we don't need to set the default provider here twice; the second one is redundant.
```suggestion
  { providerType = 'infura' } = {},
```

mcmire replied:
This has been fixed. See: [66aa9c1](https://github.com/MetaMask/metamask-extension/pull/16337/commits/66aa9c1b1c4961ba571948aea065608cb7517087)

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -95,58 +121,69 @@ export function testsForRpcMethodNotHandledByMiddleware(
-  it('does not hit Infura more than once for identical requests', async () => {
+export function testsForRpcMethodAssumingNoBlockParam(
+  method,
+  { providerType = 'infura' } = { providerType: 'infura' },
```
Nit:
```suggestion
  { providerType = 'infura' } = {},
```

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/helpers.js:
```diff
@@ -275,25 +282,34 @@ export async function withMockedInfuraCommunications(...args) {
+ * you to interact with the client via a couple of methods on that object.
  * @returns {Promise<any>} The return value of the given function.
  */
-export async function withInfuraClient(...args) {
```
Nit: I don't think it makes sense anymore for this function to allow the options to be unset. The provider type is always passed in in practice, and it seems better to require that going forward too. That also lets us simplify this function.

Instead of taking "args" and splitting it into options and callback by length, we can always expect options and a callback. This makes the types and JSDoc descriptions simpler too.

mcmire replied:
Agreed! I haven't made this change yet, so it still needs to be done.

Gudahtt commented on app/scripts/controllers/network/createInfuraClient.test.js:
```diff
@@ -1,389 +0,0 @@
-
-    describe('eth_getUncleByBlockNumberAndIndex', () => {
-      // NOTE: eth_getUncleByBlockNumberAndIndex does take a block param at the
-      // 0th index, but this is not handled by our cache middleware currently
```
Nit: this comment was dropped

mcmire replied:
It looks like this was moved to `shared-tests.js` (under "methods that assume there is no block param").

Gudahtt replied:
That is a different comment. Notice that this method is not listed under that shared test.

mcmire replied:
Fixed in[756e87d](https://github.com/MetaMask/metamask-extension/pull/16337/commits/756e87dd08aef9d8f75323faef496b8e0751e6ab).

Gudahtt commented on app/scripts/controllers/network/createInfuraClient.test.js:
```diff
@@ -1,389 +0,0 @@
-    });
-
-    describe('eth_getUncleCountByBlockNumber', () => {
-      // NOTE: eth_getUncleCountByBlockNumber does take a block param at the 0th
```
Nit: this comment was dropped

mcmire replied:
It looks like this was moved to `shared-tests.js` (under "methods that assume there is no block param").

Gudahtt replied:
That is a different comment. This method is not listed in the comment under that shared test.

mcmire replied:
Fixed in[756e87d](https://github.com/MetaMask/metamask-extension/pull/16337/commits/756e87dd08aef9d8f75323faef496b8e0751e6ab).

Gudahtt replied:
Thanks!

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/createClient.test.js:
```diff
@@ -0,0 +1,180 @@
+      );
+    });
+
+    it(`net_version does not hit ${providerType} provider, instead returning the networkId that maps to the rpc network, as a decimal string`, async () => {
```
This description isn't accurate for the RPC client, which does hit the network to answer this query.

mcmire replied:
I ended up moving this to the tests for Infura client. See: [66aa9c1](https://github.com/MetaMask/metamask-extension/pull/16337/commits/66aa9c1b1c4961ba571948aea065608cb7517087)

Gudahtt reviewed (commented):
Just finished my first complete pass of this PR! Overall this looks great, I'm finding the tests easier to follow with similar methods grouped together like that in the client tests. The changes to the helper functions look reasonable too.

I've made a few suggestions on improving readability, and found a few small things that seem to have been removed accidentally. It would also be great to remove the changes to the code-under-test in this PR, to make this easier to review and to make the commit history more readable.

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -674,61 +787,70 @@ export function testsForRpcMethodSupportingBlockParam(
-        { method, params },
-      ];
-      const mockResults = ['first result', 'second result'];
+    if (providerType === 'infura') {
```
It looks like this is exposing a genuine bug in our custom RPC provider, specifically in the cases where the block tag is unset or set to `latest`.

Gudahtt replied:
cc @mcmire, it looks like if I replace `BlockRefRewriteMiddleware` with `BlockRefMiddleware` in the network client, this bug is fixed. I'm guessing what's happening here is that `BlockRefRewriteMiddleware` is replacing `latest` with the _old_ block number without first checking the block tracker. Then the next middleware, the block cache middleware, is returning the stale cache result.

mcmire replied:
Hmm, interesting. Okay I'll take a closer look at this when I get a chance.

Gudahtt replied:
I can create a PR to fix this bug in the custom RPC client. I just tagged you because of how this relates to work you've done in the `eth-json-rpc-middleware` package. I remember you had at one point suggested that we get rid of the "child request" in `BlockRefMiddleware`, but I suspect doing that would have introduced this bug in that middleware as well.

If that is the case, we should add a test to `eth-json-rpc-middleware` to capture this behavior and delete the broken `BlockRefRewriteMiddleware` from the package completely.

Gudahtt replied:
Nevermind. This was a mistake in the tests. The `params` are being shared between the two calls here, so when that object is directly mutated in one it gets changed in the second call as well.

The middleware itself is working just fine. They both correctly ask the block tracker for the latest block number.

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -1108,31 +1166,144 @@ export function testsForRpcMethodSupportingBlockParam(
             });
+          }
+
+          it(`causes a request to fail with a custom error if an "${errorMessagePrefix}" error is thrown while making the request to Infura 5 times in a row`, async () => {
```
Another reference to Infura here that should be using `providerType`

mcmire replied:
Fixed in [66aa9c1](https://github.com/MetaMask/metamask-extension/pull/16337/commits/66aa9c1b1c4961ba571948aea065608cb7517087).

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -1257,138 +1436,161 @@ export function testsForRpcMethodSupportingBlockParam(
+        );
       });
 
       it('makes an additional request to Infura if the given block number is less than the latest block number', async () => {
```
(last one) Another reference to Infura here that should be using `providerType`

mcmire replied:
Fixed in [66aa9c1](https://github.com/MetaMask/metamask-extension/pull/16337/commits/66aa9c1b1c4961ba571948aea065608cb7517087).

carlexos commented on app/scripts/controllers/network/provider-api-tests/createClient.test.js:
```diff
@@ -0,0 +1,216 @@
+      });
+    });
+
+    describe('net_version and eth_chainId', () => {
```
Leave a reaply

mcmire committed:
Make a series of changes

* Split `createClient.test.js` into `createInfuraClient.test` and
  `createJsonRpcClient.test.js`. This way we can add Infura-specific
  tests in the future.
* Clear up some of the conditions in `shared-tests.js`, especially
  around error handling for certain HTTP status codes. It was getting a
  little difficult to read, so this should make it more straightforward.
* Correct and/or fill in some of the JSDocs.
* Ensure that for functions that take `providerType`, `providerType` is
  either "infura" or "custom" (in lieu of TypeScript).
* Correct some of the test names.
* Restore `createJsonRpcClient.js`, and add a "hack" to prevent the
  `eth_estimateGas` RPC method from getting overridden.

mcmire commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -604,39 +686,70 @@ export function testsForRpcMethodsThatCheckForBlockHashInResponse(method) {
  * number or a block tag ("latest", "earliest", or "pending") and affects how
  * the method is cached.
+ *
+ * @param options - the options
```
This has been added.

mcmire commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -340,88 +389,102 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
-        const request = { method };
+    if (
+      providerType === 'infura' ||
+      ['ECONNRESET', 'SyntaxError'].includes(errorMessagePrefix) === false
```
I ended up unrolling this loop. See: [66aa9c1](https://github.com/MetaMask/metamask-extension/pull/16337/commits/66aa9c1b1c4961ba571948aea065608cb7517087)

mcmire commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -1257,138 +1436,161 @@ export function testsForRpcMethodSupportingBlockParam(
+        );
       });
 
       it('makes an additional request to Infura if the given block number matches the latest block number', async () => {
```
Fixed in [66aa9c1](https://github.com/MetaMask/metamask-extension/pull/16337/commits/66aa9c1b1c4961ba571948aea065608cb7517087).

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Remove unused function

mcmire committed:
Use single quotes instead of backticks for strings that lack interpolation

mcmire committed:
eth_chainId for JSON-RPC client doesn't hit Infura

mcmire committed:
Fix createJsonRpcClient to allow whether test middleware is added

It turns out we have to update `createJsonRpcClient.js` in this PR,
because as soon as the file is read, `process.env.IN_TEST` is used to
determine whether `eth_estimateGas` is amended to run more slowly.
We don't want to do this for our tests.

mcmire committed:
Test JSON-RPC client handling 418 as a rate limit error instead of 429

mcmire committed:
Ensure retriable errors are tested correctly on both sides

For retriable errors that the Infura client handles which the custom RPC
client does not handle, we need to test both sides — that is, we need to
test that Infura handles them and the custom RPC client does not. And
for retriable errors that the custom RPC client handles that Infura
doesn't, we need to test that on both sides too.

mcmire commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -750,100 +2176,144 @@ export function testsForRpcMethodSupportingBlockParam(
       });
-    });
 
-    if (blockParamType === 'none') {
```
I noticed this `if` statement while perusing this file. I think I left this `if` statement in a previous PR for debugging purposes, but this should not be here — or rather, this code should not be wrapped in an `if` statement. I've unwrapped the code in this PR, but have also extracted it to another PR: https://github.com/MetaMask/metamask-extension/pull/16906

mcmire commented on app/scripts/controllers/network/createInfuraClient.test.js:
```diff
@@ -1,389 +0,0 @@
-      });
-    });
-
-    describe('eth_gasPrice', () => {
```
Agreed. I haven't added this test back yet.

mcmire replied:
Added back in [8a1e46b](https://github.com/MetaMask/metamask-extension/pull/16337/commits/8a1e46b1dcda4b0308ec01991ec7c359a43c7470).

mcmire commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -340,88 +389,102 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
-      await withMockedInfuraCommunications(async (comms) => {
-        const request = { method };
+    if (
+      providerType === 'infura' ||
```
I think it makes sense to add a little something for `ECONNRESET` (even if it might be a good idea to place this knowledge eventually live in `eth-json-rpc-infura`). As for `SyntaxError`, I'm not sure that it is unique to the Infura middleware. It looks like the fetch middleware handles this already, at least in spirit. It [attempts to parse the response body as JSON](https://github.com/MetaMask/eth-json-rpc-middleware/blob/254a66cf5f067f35db96c6a5740c8da9a79e495d/src/fetch.ts#L64-L70), and when that fails (due to a SyntaxError) it will throw an error which is then picked up as a retriable error.

mcmire replied:
Added comment for `ECONNRESET` in [f676e08](https://github.com/MetaMask/metamask-extension/pull/16337/commits/f676e08e27ce08e82ee780eb2712bb792e45735b), will add more comments.

Gudahtt replied:
Great, thanks!

mcmire commented:
@BelfordZ Sorry if I am stepping on your toes here, but I wanted to see if we could tie the loose ends on this PR. I took the time to address Mark's feedback above. In addition, there was some behavior that the custom JSON-RPC client has which the Infura client does not, and I made sure that was tested. Finally, I made sure that any behavior one client had, we were testing that this behavior was not applicable to the other client. For instance, the custom JSON-RPC client treats 418 as a rate limit error, but the Infura client treats 429 as a rate limit error. I made sure that we test that 418 is not treated as special by the Infura client and that 429 is not treated as special by the custom JSON-RPC client. In the process of doing this I had to unbundle some of the tests. In other words, instead of deciding the result of a request inside of a test based on the provider, I duplicated the test, then said:

```
if (providerType === 'infura') {
  // (test for Infura goes here)
} else {
  // (test for custom RPC-client goes here)
}
```

This allowed me to name the tests differently to describe the different behavior. I realize this may have not been your plan but hopefully you can see why I did that.

I'm mostly off tomorrow (Monday) but if you want to chat about this I'll be available again on Tuesday.

mcmire committed:
Add comment explaining why ECONNRESET is specially handled for Infura

mcmire committed:
Add missing comments back

mcmire committed:
Add eth_gasPrice back

mcmire committed:
Relocate net_version and eth_chainId tests

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Simplify withClient (now withNetworkClient) and withMockedCommunications

mcmire committed:
Document more of the differences between tests

mcmire committed:
Remove TODOs

mcmire commented:
@Gudahtt I believe I've addressed all of your comments but let me know if you see anything else.

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Pull these tests out

mcmire committed:
Little tweaks

mcmire commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -54,33 +97,311 @@ async function waitForPromiseToBeFulfilledAfterRunningAllTimers(
+
+    describe('other methods', () => {
+      describe('eth_getTransactionByHash', () => {
+        it("refreshes the block tracker's current block if it is less than the block number that comes back in the response", async () => {
```
@BelfordZ You had placed these in `testsForRpcMethodsThatCheckForBlockHashInResponse`, but I don't think that's the right place for these tests. The purpose of `testsForRpcMethodsThatCheckForBlockHashInResponse` was to test for behavior related to caching (specifically [these lines in the `block-cache` middleware](https://github.com/MetaMask/eth-json-rpc-middleware/blob/254a66cf5f067f35db96c6a5740c8da9a79e495d/src/block-cache.ts#L110-L128) and doesn't have anything to do with `blockHash` being in the `result`. 

This raises a question of organization and whether the way you've chosen is the best approach. Right now you're grouping RPC methods together by which `testsFor*` function is being called. However, this is a bit rigid because it means if you want to add a test for an RPC method you need to repeat the `describe` for that RPC method and possibly place it in some other group. If we instead organize primarily by RPC method, then we are free to place whatever tests we like in the `describe` for that RPC method. So it looks like:

``` javascript
describe('RPC method 1', () => {
  it('test 1', () => {
    // ...
  });

  it('test 2', () => {
    // ...
  });

  it('test 3', () => {
    // ...
  });
})

describe('RPC method 2', () => {
  it('test 1', () => {
    // ...
  });

  it('test 2', () => {
    // ...
  });

  it('test 3', () => {
    // ...
  });
})
```

The `testsFor*` functions are just shorthands for defining a bunch of related tests in one go but at the end of the day you would still be able think of it like the above.

I realize that the reason why you might have to decided to reorganize the tests this way is due to the naming of the `testsFor*` functions. I should come up with better names that explain the behavior being tested by the tests, which they don't do a good job of right now. I will do this in a separate PR, but just wanted to make a note of it in the meantime.

mcmire replied:
We may introduce some changes in a future PR based on work Zach did locally that didn't make it into this PR.

mcmire commented:
Synced with @BelfordZ this morning on this PR. I will take this over and get this merged.

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Fix test

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire commented on app/scripts/controllers/network/createJsonRpcClient.js:
```diff
@@ -10,22 +10,22 @@ import {
     ...blockTrackerOpts,
     provider: blockProvider,
   });
+  const testMiddlewares = process.env.IN_TEST
```
For some reason, the test middleware added here is interfering with the running of the tests for `createJsonRpcClient`. I'm not sure whether the middleware is set up correctly or not, but after the promise resolves in `createEstimateGasDelayTestMiddleware`, the middleware after this point is not getting run, so requests through the provider are never completing. 

To get around this in a minimal fashion, we want to disable this middleware only for these network client tests. The problem is that previously we were reading `process.env.IN_TEST` when this file was 
read. This made it impossible to prevent this test middleware from getting added by unsetting that environment variable. By checking for the variable inside of `createJsonRpcClient`, we can unset that variable before initializing the network client in the test helper and then reset it afterward.

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -339,41 +769,47 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
-    it(`retries the request to Infura up to 5 times if an "${errorMessagePrefix}" error is thrown while making the request, returning the successful result if there is one on the 5th try`, async () => {
-      await withMockedInfuraCommunications(async (comms) => {
+  // Both the Infura and fetch middleware detect ETIMEDOUT errors and will
+  // automatically retry the request to the RPC endpoint in question, but both
```
Nit: could we move the "success on 5th try" case outside of this `if` block, if the behavior is the same?

Gudahtt replied:
I see this applies to a few other cases as well (the other `ETIMEDOUT` test, and the 503/504 handling upon 5 successive failures).

mcmire replied:
Yes absolutely. I can make this change today.

mcmire replied:
Fixed in [6898d7e](https://github.com/MetaMask/metamask-extension/pull/16337/commits/6898d7e823a0a3db2b9869a583ae8f6a2aa458e5).

Gudahtt reviewed (dismissed):
LGTM!

mcmire committed:
No need to duplicate tests if they're the same between Infura and non-Infura

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire commented:
@BelfordZ and I will do a final pass on this and approve this together.

mcmire committed:
Remove 'jest-environment node'

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

mcmire committed:
Simplify these tests

mcmire committed:
Remove extra mock result

mcmire committed:
Merge branch 'develop' into fix/network-test-json-rpc-client

BelfordZ commented:
Reviewed Elliots updates -- looks great, nice work. 

LGTM! :heavy_check_mark:
